WorldPainter/WPCore/src/main/java/org/pepsoft/worldpainter/layers/bo2/Structure.java

```diff
@@ -182,18 +214,20 @@ public static Structure load(String objectName, InputStream inputStream) throws
 
         // Remove palette, blocks and entities from the tag so we don't waste space
         root.setTag(TAG_PALETTE_, null);
+        root.setTag(TAG_PALETTES, null);
         root.setTag(TAG_BLOCKS_, null);
         root.setTag(TAG_ENTITIES_, null);
 
-        return new Structure(root, objectName, blocks, (! entities.isEmpty()) ? ImmutableList.copyOf(entities) : null, (! tileEntities.isEmpty()) ? ImmutableList.copyOf(tileEntities) : null);
+        return new Structure(root, objectName, blocks, (! entities.isEmpty()) ? ImmutableList.copyOf(entities) : null, (! tileEntities.isEmpty()) ? ImmutableMap.copyOf(tileEntities) : null);
     }
 
     private final CompoundTag root;
-    private final Map<Point3i, Material> blocks;
+    private final Map<Integer, Map<Point3i, Material>> blocks;
```

Comment:
To make this change backwards compatible, so that existing `Structure`s can also still be loaded, Java serialization allows a pretty elegant mechanism. What you would do is keep both the old and the new fields, and add a `readObject()` method, which would first call `in.defaultReadObject()` (which would perform the default deserialization process), and then check whether the new fields are set, and if not, migrate the old fields to the new. For examples and inspiration look at e.g. `World.readObject()` or any of the other implementations of `readObject()` in the codebase (although I think this situation is simple enough that it's not yet necessary to introduce a `version` field, which many of the other implementations do).